kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js

```diff
@@ -0,0 +1,323 @@
+
+    // Constants
+    var t = html.tag,
+        div = t('div'),
+        span = t('span'),
+        b = t('b'),
+        select = t('select'),
+        option = t('option');
+
+    function factory(config) {
+        var spec = config.parameterSpec,
+            parent,
+            container,
+            runtime = Runtime.make(),
+            bus = runtime.bus().connect(),
+            channel = bus.channel(config.channelName),
+            ui,
+            dataSource = 'ftp_staging', // only option now, should come from the parameterSpec.
+            model = {
+                value: undefined
+            },
+            fileStaging = new FileStaging(runtime.config('services.ftp_api_url.url'),
+                runtime.userId(), {token: runtime.authToken()}),
+            eventListeners = [];
+
+        function makeInputControl() {
+            var selectOptions;
+            var selectElem = select({
+                class: 'form-control',
+                dataElement: 'input',
+                style: {
+                    width: '100%'
+                },
+                id: html.genId()
+            }, [option({ value: '' }, '')].concat(selectOptions));
+
+            return selectElem;
+        }
+
+        // CONTROL
+        function getControlValue() {
+            var control = ui.getElement('input-container.input'),
+                selected = control.selectedOptions;
+            if (selected.length === 0) {
+                return;
+            }
+            // we are modeling a single string value, so we always just get the
+            // first selected element, which is all there should be!
+            return selected.item(0).value;
+        }
+
+        /**
+         * Sets the dropdown value to the given value. Constructs an id from it that (should)
+         * be unique enough to apply to the dropdown.
+         */
+        function setControlValue(value) {
+            var control = ui.getElement('input-container.input');
+            if ($(control).find('option[value="' + value + '"]').length) {
+                $(control).val(value).trigger('change');
+            } else {
+                var newOption = new Option(value, value, true, true);
+                $(control).append(newOption).trigger('change');
+            }
+        }
+
+        // MODEL
+
+        function setModelValue(value) {
+            if (model.value === undefined) {
+                return;
+            }
+            if (model.value !== value) {
+                model.value = value;
+            }
+        }
+
+        function resetModelValue() {
+            setModelValue(spec.data.defaultValue);
+        }
+
+        function getModelValue() {
+            return model.value;
+        }
+
+        // VALIDATION
+
+        function validate() {
+            return Promise.try(function() {
+                var selectedItem = getControlValue(),
+                    validationConstraints = {
+                        min_length: 1,
```

Comment:
shouldn't the validation constraints be driven from the spec or sdk.js? trawling back through memory of how this is used elsewhere.

Reply:
also in theory wouldn't the validation go against the data source? I mean, in actual usage the value is derived from a dropdown derived from the data source, but as far as the control knows, at some future time this may not be true. That is a problem in general for this control, though, since all files will eventually disappear and thus every instance of this parameter will become invalid. Hmm..

Reply:
Yes, but these aren't expected to have string constraints. Those are just there as default values, though sdk.js should cough up the defaults itself, instead of the input widget. I'll change that.